### PR TITLE
Allow modifications to variants of components for publishing

### PR DESCRIPTION
**Pull request by jjohannes**
Implementation of #10428

### CODE REVIEW

File: subprojects/docs/src/docs/userguide/dep-man/06-publishing/publishing.adoc [hunk 1]
```diff
@@ -265,16 +265,17 @@ include::sample[dir="userguide/dependencyManagement/modelingFeatures/crossProjec
 include::sample[dir="userguide/dependencyManagement/modelingFeatures/crossProjectPublications/advanced-published/kotlin/buildSrc/src/main/kotlin/com/acme",files="InstrumentedJarsPlugin.kt[tags=create_adhoc_component]"]
 ====
 
-The `AdhocComponentWithVariants` component declares a method called `addVariantsFromConfiguration` which accepts two parameters:
+The `AdhocComponentWithVariants` component declares two method called `addVariantsFromConfiguration` and `withVariantsFromConfiguration` which accept two parameters:
```
Comment (melix) [inline subprojects/docs/src/docs/userguide/dep-man/06-publishing/publishing.adoc]: :x: Two method**s**

File: subprojects/docs/src/docs/userguide/java_testing.adoc [hunk 2]
```diff
@@ -570,3 +570,7 @@ include::{samplesPath}/userguide/java/fixtures/dependencyInsight.out[]
 The error message mentions the missing `com.google.code.gson:gson-test-fixtures` capability, which is indeed not defined for this library.
 That's because by convention, for projects that use the `java-test-fixtures` plugin, Gradle automatically creates test fixtures variants with a capability whose name is the name of the main component, with the appendix `-test-fixtures`.
 
+[NOTE]
+====
+If you publish your library and use test fixtures, but do not want to publish the fixtures, you can deactivate publishing of the _test fixtures variants_ using `components.java.withVariantsFromConfiguration(configurations.testFixturesApiElements) { skip() }` and `components.java.withVariantsFromConfiguration(configurations.testFixturesRuntimeElements) { skip() }`.
```
Comment (melix) [inline subprojects/docs/src/docs/userguide/java_testing.adoc]: :o: Ideally this should be tested in samples with the Kotlin DSL too.

Comment (jjohannes) [reply subprojects/docs/src/docs/userguide/java_testing.adoc]: Good point. Test coverage and it's easier for users to _copypaste_ the solution for the DSL they use.
I added the sample.

File: subprojects/docs/src/docs/userguide/dep-man/06-publishing/publishing.adoc (hunk 1 above)
Comment (jjohannes) [reply subprojects/docs/src/docs/userguide/dep-man/06-publishing/publishing.adoc]: fixed

**Comment by jjohannes** [comment]
@melix addressed you feedback. Thanks.

> Should you leverage this to also fail publication if the resulting set of filtered variants is empty?

I think this should be a follow up and could be part of #10736. I added a comment to that issue.

**Pull request merged**